### PR TITLE
[TG-1940] Fix Java ternary-compare against NaN

### PR DESCRIPTION
**Pull request by smowton**
This was supposed to check if either operand is NaN and respond appropriately,
but using float-eq to do the test was incorrect, as NaN == x is false for any x,
whether NaN or not. This patch uses the isnan exprt to do the test instead, and adds
a regression test.

### CODE REVIEW

**Comment by smowton** [comment]
@allredj @antlechner @peterschrammel here's the Java NaN fix.

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 1]
```diff
@@ -2062,8 +2062,8 @@ codet java_bytecode_convert_methodt::convert_instructions(
       // (value1 == NaN || value2 == NaN) ?
       //   nan_value : value1 == value2 ? 0  : value1 < value2 -1 ? 1 : 0;
 
-      exprt nan_op0=ieee_float_equal_exprt(nan_expr, op[0]);
-      exprt nan_op1=ieee_float_equal_exprt(nan_expr, op[1]);
+      isnan_exprt nan_op0(op[0]);
```
Comment (mgudemann) [inline src/java_bytecode/java_bytecode_convert_method.cpp:2060]: please also remove `      const constant_exprt nan_expr(nan.to_expr());` as it's not used any more

**Review by mgudemann — APPROVED** [review]
LGTM, please remove the now unused `nan_expr`

**Review by allredj — APPROVED** [review]
Looks better than the original version. Would it make sense to add tests for `double`, or does this all fall into the same bucket?

**Review by martin-cs — APPROVED** [review]
Yes. Please.

**Comment by smowton** [comment]
@mgudemann @allredj changes applied, will merge once passed CI

**Pull request merged**